File: src/CHTL/CHTLLexer/Lexer.cpp
#include "Lexer.h"
#include <cctype>

namespace CHTL {

Lexer::Lexer(const std::string& input) : m_input(input) {
    readChar();
}

void Lexer::readChar() {
    if (m_readPosition >= m_input.length()) {
        m_char = 0; // EOF
    } else {
        m_char = m_input[m_readPosition];
    }
    m_position = m_readPosition;
    m_readPosition++;
    m_column++;
    if (m_char == '\n') {
        m_line++;
        m_column = 0;
    }
}

char Lexer::peekChar() const {
    if (m_readPosition >= m_input.length()) {
        return 0;
    }
    return m_input[m_readPosition];
}

void Lexer::skipWhitespace() {
    while (m_char == ' ' || m_char == '\t' || m_char == '\n' || m_char == '\r') {
        readChar();
    }
}

void Lexer::skipSingleLineComment() {
    while (m_char != '\n' && m_char != 0) {
        readChar();
    }
}

void Lexer::skipMultiLineComment() {
    while (true) {
        if (m_char == 0) {
            // Unterminated multi-line comment
            break;
        }
        if (m_char == '*' && peekChar() == '/') {
            readChar(); // consume '*'
            readChar(); // consume '/'
            break;
        }
        readChar();
    }
}

Token Lexer::readIdentifier() {
    int start_pos = m_position;
    while (isalpha(m_char) || m_char == '_') {
        readChar();
    }
    return {TokenType::IDENTIFIER, m_input.substr(start_pos, m_position - start_pos), m_line, m_column};
}

Token Lexer::readString(char quote) {
    int start_pos = m_position + 1;
    readChar(); // consume opening quote
    while (m_char != quote && m_char != 0) {
        readChar();
    }
    std::string literal = m_input.substr(start_pos, m_position - start_pos);
    readChar(); // consume closing quote
    return {TokenType::STRING, literal, m_line, m_column};
}

Token Lexer::readUnquotedLiteral() {
    int start_pos = m_position;
    // Unquoted literals can contain almost anything until a delimiter.
    // This is a simplified rule for now. The parser will provide more context.
    while (m_char != ';' && m_char != '{' && m_char != '}' && m_char != '\n' && m_char != 0) {
        readChar();
    }
    // Trim trailing whitespace from the literal
    size_t end_pos = m_position;
    while (end_pos > start_pos && isspace(m_input[end_pos - 1])) {
        end_pos--;
    }
    return {TokenType::STRING, m_input.substr(start_pos, end_pos - start_pos), m_line, m_column};
}

Token Lexer::nextToken() {
    skipWhitespace();

    Token tok;
    tok.line = m_line;
    tok.column = m_column;

    switch (m_char) {
        case '{': tok = {TokenType::LBRACE, "{", m_line, m_column}; break;
        case '}': tok = {TokenType::RBRACE, "}", m_line, m_column}; break;
        case '(': tok = {TokenType::LPAREN, "(", m_line, m_column}; break;
        case ')': tok = {TokenType::RPAREN, ")", m_line, m_column}; break;
        case '[': tok = {TokenType::LBRACKET, "[", m_line, m_column}; break;
        case ']': tok = {TokenType::RBRACKET, "]", m_line, m_column}; break;
        case ':': tok = {TokenType::COLON, ":", m_line, m_column}; break;
        case '=': tok = {TokenType::EQUALS, "=", m_line, m_column}; break;
        case ';': tok = {TokenType::SEMICOLON, ";", m_line, m_column}; break;
        case ',': tok = {TokenType::COMMA, ",", m_line, m_column}; break;
        case '.': tok = {TokenType::DOT, ".", m_line, m_column}; break;
        case '&': tok = {TokenType::AMPERSAND, "&", m_line, m_column}; break;
        case '?': tok = {TokenType::QUESTION, "?", m_line, m_column}; break;
        case '+': tok = {TokenType::PLUS, "+", m_line, m_column}; break;
        case '*': tok = {TokenType::STAR, "*", m_line, m_column}; break;
        case '%': tok = {TokenType::PERCENT, "%", m_line, m_column}; break;
        case '"': return readString('"');
        case '\'': return readString('\'');
        case '-':
            if (peekChar() == '-') {
                readChar(); // consume first '-'
                readChar(); // consume second '-'
                int start_pos = m_position;
                skipSingleLineComment();
                tok = {TokenType::COMMENT_GEN, m_input.substr(start_pos, m_position - start_pos), m_line, m_column};
                // Don't call readChar() at the end for this one
                return tok;
            } else {
                tok = {TokenType::MINUS, "-", m_line, m_column};
            }
            break;
        case '/':
            if (peekChar() == '/') {
                skipSingleLineComment();
                return nextToken(); // Get next token after comment
            } else if (peekChar() == '*') {
                readChar(); // consume '/'
                readChar(); // consume '*'
                skipMultiLineComment();
                return nextToken(); // Get next token after comment
            } else {
                tok = {TokenType::SLASH, "/", m_line, m_column};
            }
            break;
        case 0:
            tok = {TokenType::END_OF_FILE, "", m_line, m_column};
            break;
        default:
            if (isalpha(m_char) || m_char == '_') {
                // This is a simplification. Keywords vs. identifiers vs. unquoted strings
                // will need more robust handling. For now, we assume alpha-starting things are identifiers.
                return readIdentifier();
            } else {
                // If it's not a recognized symbol or an identifier, it could be an unquoted literal.
                // This is a basic implementation. We'll refine this.
                if (isprint(m_char)) {
                     return readUnquotedLiteral();
                }
                tok = {TokenType::ILLEGAL, std::string(1, m_char), m_line, m_column};
            }
            break;
    }

    readChar();
    return tok;
}

} // namespace CHTL

File: chtl/src/chtl_js_compiler/lexer/lexer.rs
use super::token::Token;

pub struct Lexer<'a> {
    input: &'a str,
    position: usize,
    read_position: usize,
    ch: u8,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        let mut l = Lexer {
            input,
            position: 0,
            read_position: 0,
            ch: 0,
        };
        l.read_char();
        l
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() {
            self.ch = 0;
        } else {
            self.ch = self.input.as_bytes()[self.read_position];
        }
        self.position = self.read_position;
        self.read_position += 1;
    }

    fn peek_char(&self) -> u8 {
        if self.read_position >= self.input.len() {
            0
        } else {
            self.input.as_bytes()[self.read_position]
        }
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.read_char();
        }
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();

        let tok = match self.ch {
            b'{' => {
                if self.peek_char() == b'{' {
                    self.read_char(); // consume first {
                    self.read_char(); // consume second {
                    let content = self.read_selector_content();
                    return Token::SelectorContent(content);
                }
                Token::OpenBrace
            }
            b'(' => Token::OpenParen,
            b')' => Token::CloseParen,
            b',' => Token::Comma,
            b';' => Token::Semicolon,
            b':' => Token::Colon,
            b'.' => Token::Dot,
            b'-' => {
                if self.peek_char() == b'>' {
                    self.read_char();
                    return Token::Arrow;
                }
                Token::Illegal("-".to_string())
            }
            b'=' => {
                if self.peek_char() == b'>' {
                    self.read_char();
                    return Token::FatArrow;
                }
                Token::Illegal("=".to_string())
            }
            b'"' | b'\'' => return self.read_string(self.ch),
            _ => {
                if self.ch.is_ascii_alphabetic() || self.ch == b'_' {
                    let ident = self.read_identifier();
                    return Token::Identifier(ident);
                } else if self.ch == 0 {
                    return Token::EndOfFile;
                }
                Token::Illegal(self.ch.to_string())
            }
        };

        self.read_char();
        tok
    }

    fn read_identifier(&mut self) -> String {
        let position = self.position;
        while self.ch.is_ascii_alphanumeric() || self.ch == b'_' {
            self.read_char();
        }
        self.input[position..self.position].to_string()
    }

    fn read_string(&mut self, quote_type: u8) -> Token {
        let position = self.position + 1;
        loop {
            self.read_char();
            if self.ch == quote_type || self.ch == 0 {
                break;
            }
        }
        if self.ch == 0 {
            Token::Illegal("Unterminated string".to_string())
        } else {
            // consume closing quote
            self.read_char();
            Token::StringLiteral(self.input[position..self.position - 1].to_string())
        }
    }

    fn read_selector_content(&mut self) -> String {
        let position = self.position;
        while !(self.ch == b'}' && self.peek_char() == b'}') && self.ch != 0 {
            self.read_char();
        }
        let content = self.input[position..self.position].trim().to_string();
        if self.ch != 0 {
            self.read_char(); // consume first }
            self.read_char(); // consume second }
        }
        content
    }
}
